src/graph.rs

```diff
@@ -0,0 +1,343 @@
+                    )?;
+                    for (sink, edge) in &node.outgoing_edges {
+                        write!(
+                            f,
+                            "edge {} -> {}\n{}",
+                            node_index,
+                            *sink,
+                            edge.attributes.display_with(ctx, graph)
+                        )?;
+                    }
+                }
+                Ok(())
+            }
+        }
+
+        DisplayGraph(self, ctx)
+    }
+}
+
+impl<'tree> Index<SyntaxNodeRef> for Graph<'tree> {
+    type Output = Node<'tree>;
+    fn index(&self, index: SyntaxNodeRef) -> &Node<'tree> {
+        &self.syntax_nodes[&index]
+    }
+}
+
+impl Index<GraphNodeRef> for Graph<'_> {
+    type Output = GraphNode;
+    fn index(&self, index: GraphNodeRef) -> &GraphNode {
+        &self.graph_nodes[index.0 as usize]
+    }
+}
+
+impl<'tree> IndexMut<GraphNodeRef> for Graph<'_> {
+    fn index_mut(&mut self, index: GraphNodeRef) -> &mut GraphNode {
+        &mut self.graph_nodes[index.0 as usize]
+    }
+}
+
+/// A node in a graph
+pub struct GraphNode {
+    outgoing_edges: SmallVec<[(GraphNodeID, Edge); 8]>,
+    /// The set of attributes associated with this graph node
+    pub attributes: Attributes,
+}
+
+impl GraphNode {
+    fn new() -> GraphNode {
+        GraphNode {
+            outgoing_edges: SmallVec::new(),
+            attributes: Attributes::new(),
+        }
+    }
+
+    /// Adds an edge to this node.  There can be at most one edge connecting any two graph nodes;
+    /// the result indicates whether the edge is new (`Ok`) or already existed (`Err`).  In either
+    /// case, you also get a mutable reference to the [`Edge`][] instance for the edge.
+    pub fn add_edge(&mut self, sink: GraphNodeRef) -> Result<&mut Edge, &mut Edge> {
+        let sink = sink.0;
+        match self
+            .outgoing_edges
+            .binary_search_by_key(&sink, |(sink, _)| *sink)
+        {
+            Ok(index) => Err(&mut self.outgoing_edges[index].1),
+            Err(index) => {
+                self.outgoing_edges.insert(index, (sink, Edge::new()));
+                Ok(&mut self.outgoing_edges[index].1)
+            }
+        }
+    }
+
+    /// Returns a reference to an outgoing edge from this node, if it exists.
+    pub fn get_edge(&self, sink: GraphNodeRef) -> Option<&Edge> {
+        let sink = sink.0;
+        self.outgoing_edges
+            .binary_search_by_key(&sink, |(sink, _)| *sink)
+            .ok()
+            .map(|index| &self.outgoing_edges[index].1)
+    }
+
+    /// Returns a mutable reference to an outgoing edge from this node, if it exists.
+    pub fn get_edge_mut(&mut self, sink: GraphNodeRef) -> Option<&mut Edge> {
+        let sink = sink.0;
+        self.outgoing_edges
+            .binary_search_by_key(&sink, |(sink, _)| *sink)
+            .ok()
+            .map(move |index| &mut self.outgoing_edges[index].1)
+    }
```

Comment:
Likewise here, is there a convenient means of collapsing the two definitions to avoid duplicating the binary search? If we can’t define one in terms of the other, could we internalize the iteration by passing in a closure (ideally non-escaping so it can be stack-allocated &c.) or something like that?

Reply:
There's no way in Rust to generalize across `mut`/not `mut`, unfortunately.  For me, these method are small enough and near enough to each other that the code duplication isn't something that I worry about.  `binary_search_by_key` is a common enough part of the stdlib that I think it's a readable idiom.